wrappers/core18.go

```diff
@@ -141,16 +141,49 @@ type AddSnapdSnapServicesOptions struct {
 	Preseeding bool
 }
 
+type SnapdRestart struct {
+	Sysd systemd.Systemd
+}
+
+func (r *SnapdRestart) Restart() error {
+	if err := r.Sysd.StartNoBlock([]string{"snapd.apparmor.service"}); err != nil {
+		return err
+	}
+
+	// and finally start snapd.service (it will stop by itself and gets
+	// started by systemd then)
+	// Because of the file lock held on the snapstate by the Overlord, the new
+	// snapd will block there until we release it. For this reason, we cannot
+	// start the unit in blocking mode.
+	// TODO: move/share this responsibility with daemon so that we can make the
+	// start blocking again
+	if err := r.Sysd.StartNoBlock([]string{"snapd.service"}); err != nil {
+		return err
+	}
```

Comment:
after this point I agree is a bit less clear, as the new snapd will hopefully be tried to start?